File: haskell/meetup/Meetup.hs
module Meetup (Weekday(..), Schedule(..), meetupDay) where

import Data.Time.Calendar (fromGregorian, Day, gregorianMonthLength)
import Data.Time.Calendar.WeekDate (toWeekDate)

data Weekday = Monday
             | Tuesday
             | Wednesday
             | Thursday
             | Friday
             | Saturday
             | Sunday
  deriving Enum

data Schedule = First
              | Second
              | Third
              | Fourth
              | Teenth
              | Last

meetupDay :: Schedule -> Weekday -> Integer -> Int -> Day
meetupDay schedule weekday year month = case schedule of
  First  -> weekdaysInMonth !! 0
  Second -> weekdaysInMonth !! 1
  Third  -> weekdaysInMonth !! 2
  Fourth -> weekdaysInMonth !! 3
  Teenth -> head (weekdays [13..19])
  Last   -> last weekdaysInMonth
  where
    weekdaysInMonth = weekdays [1..(gregorianMonthLength year month)]
    weekdays = filter (sameWeekday . toWeekDate) . dates
    sameWeekday (_, _, wd) = wd == succ (fromEnum weekday)
    dates = map (fromGregorian year month)

File: python/meetup/meetup.py
from datetime import date, timedelta
from calendar import monthrange

weekday_table = {
  "Monday": 0,
  "Tuesday": 1,
  "Wednesday": 2,
  "Thursday": 3,
  "Friday": 4,
  "Saturday": 5,
  "Sunday": 6,
}

def meetup_day(year, month, weekday, label):
  def find_weekday(start, direction):
    day = date(year, month, start)
    while day.weekday() != weekday_table[weekday]:
      day += timedelta(direction)
    return day
  days_in_month = monthrange(year, month)[1]
  firstday = find_weekday(1, 1)
  lastday = find_weekday(days_in_month, -1)
  if label == "1st":
    return firstday
  elif label == "2nd":
    return firstday + timedelta(7)
  elif label == "3rd":
    return firstday + timedelta(14)
  elif label == "4th":
    return firstday + timedelta(21)
  elif label == "5th":
    if lastday - firstday != timedelta(28): raise ValueError
    return lastday
  elif label == "last":
    return lastday
  elif label == "teenth":
    return find_weekday(13, 1)
  else:
    raise ValueError
